### DISCUSSION TITLE
vlc: cannot play copy-protected dvds (private-dev)

### DISCUSSION THREAD

**Discussion by frank-github** [netblue30/firejail#5391] [Q&A] [answered]
Hi,

Under firejail totem and vlc cannot play copy-protected DVDs.
Using distributed default configuration for totem or vlc.
Looks like vlc/totem can not find libdvdcss2/libdvdread8

If running totem/vlc without firejail, control DVDs running well.

How can I change the configurations so that libdvdcss2/libdvdread8 are found?

```
Linux debian 5.18.0-0.bpo.1-amd64 #1 SMP PREEMPT_DYNAMIC Debian 5.18.2-1~bpo11+1 (2022-06-14) x86_64 GNU/Linux
firejail version 0.9.70-1~bpo11+1
```

bye, frank

Relates to:

* #6468

**Comment by reinerh** [comment]
I think libdvdcss stores the keys in ~/.dvdcss, which is not allowed in the current profile.
Can you try adding the following to `totem.local`?
```
mkdir ${HOME}/.dvdcss
whitelist ${HOME}/.dvdcss
```

**Comment by frank-github** [comment]
Hi, that does not help.

I found these messages when starting totem in a terminal:

```
libdvdread: Attempting to use device /dev/sr1 mounted on /media/frank/HP_AND_THE_PRISONER_OF_AZKABAN for CSS authentication
libdvdread: Could not open /dev/sr1 with libdvdcss.
libdvdread: Can't open /dev/sr1 for reading
libdvdread: Device /dev/sr1 inaccessible, CSS authentication not available.
libdvdread: Attempting to use device /dev/sr1 mounted on /media/frank/HP_AND_THE_PRISONER_OF_AZKABAN for CSS authentication
libdvdread: Could not open /dev/sr1 with libdvdcss.
libdvdread: Can't open /dev/sr1 for reading
libdvdread: Device /dev/sr1 inaccessible, CSS authentication not available.
libdvdnav: Can't read name block. Probably not a DVD-ROM device.
libdvdnav: vm: dvd_read_name failed
```

Maybe some more hints.

I'm running this debian in a VM.
if I use this VM on vmhost with the internal dvd-drive, totem/vlc can play dvd, even running under firejail.
if vmhost does not have an internal dvd-drive and I use an external usb-dvd-drive, totem/vlc can not play movie with firejail, but only without.

**Reply by reinerh** [reply]
Sounds like `/dev/srX` is not available in the jail. Can you try adding `ignore private-dev` to `totem.local`?

**Reply by reinerh** [reply]
Also it looks like the device is already mounted. That shouldn't be needed when playing a DVD. No idea if this might conflict.

**Comment by frank-github** [comment]
Hey you got it :-)

With these lines for vlc and totem, dvds are running now:

```
mkdir ${HOME}/.dvdcss
whitelist ${HOME}/.dvdcss
ignore private-dev
```

Does `ignore private-dev` have any disadvantages?

**Reply by reinerh** [reply]
> do "ignore private-dev" have any disadvantage?

All devices are still available in /dev, instead of only a few selected ones.

Though I think there is also a bug in firejail. According to the firejail-profiles manpage, private-dev should still allow disc drives, and there is an additional setting `nodvd` that could be used to forbid dvd drives.
I think the naming /dev/srX for a DVD device is a bit unusual?

**Reply by rusty-snake** [reply]
Actually srX is very common to me.

**Reply by reinerh** [reply]
Ok, I have no optical drive in my laptop currently, so I wasn't able to check how they are usually named. :)

**Reply by rusty-snake** [reply]
The problem is that we only mount `sr0`. Adding `sr0..9` range, would fix it.

https://github.com/netblue30/firejail/blob/393c5beff2686d7732221dadb6730917f24835a0/src/firejail/fs_dev.c#L80

**Reply by frank-github** [reply]
> Though I think there is also a bug in firejail. According to the firejail-profiles manpage, private-dev should still allow disc drives, and there is an additional setting nodvd that could be used to forbid dvd drives.

I do not think it's a bug in general, because it works with internal dvd devices
my current laptop do not have a optical device anymore, so I have to use an external usb-device.
Only with this szenario firejail seems to have a problem

**Reply by reinerh** [reply]
> The problem is that we only mount `sr0`. Adding `sr0..9` range, would fix it.

Ah, nice finding! :+1:

**Reply by frank-github** [reply]
> The problem is that we only mount sr0. Adding sr0..9 range, would fix it.

I had wonder myself, why my external dvd device has sr**1**
at the moment it occurred to me why this is so.
my vm has a DVD device mounted by default, this is used when the VM host has an internal drive. So /dev/sr0 is already occupied.
Now when I connect the external DVD drive, this of course gets /dev/sr1

**Comment by kmk3** [comment, marked as answer]
@rusty-snake [on Oct 2, 2022](https://github.com/netblue30/firejail/discussions/5391#discussioncomment-3782950):

> The problem is that we only mount `sr0`. Adding `sr0..9` range, would fix it.
> 
> https://github.com/netblue30/firejail/blob/393c5beff2686d7732221dadb6730917f24835a0/src/firejail/fs_dev.c#L80

This is being tracked at:

* #2723

Which would probably make this use-case work with `private-dev`.

**Comment by kmk3** [comment]
@frank-github

Is `ignore private-dev` enough or do you actually need these lines for
vlc/totem?

```
mkdir ${HOME}/.dvdcss
whitelist ${HOME}/.dvdcss
```

It looks like ~/.dvdcss is just used for caching.

If it's needed, it should probably be added to vlc.profile and to other
profiles that might use use libdvdcss.

**Reply by frank-github** [reply]
Unfortunately, I can't really say for sure.
I haven't used my external DVD drive in a long time.
